### DataBase

#### 什么是数据库？
 
数据库（Database）是按照数据结构来组织、存储和管理数据的仓库。每个数据库都有一个或多个不同的 API 用于创建，访问，管理，搜索和复制所保存的数据。我们也可以将数据存储在文件中，但是在文件中读写数据速度相对较慢。因此，我们使用数据库管理系统来存储和管理大数据量。

#### 数据库的类型

关系型数据库(RDBMS)：

建立在关系模型基础上的数据库，借助于集合代数等数学概念和方法来处理数据库中的数据。主流为Mysql和Oracle数据库，互联网运维最常用的是MySQL。可通过SQL结构化查询语句存储数据，遵循ACID规则。

特点：

- 数据以表格的形式出现
- 每行为各种记录名称
- 每列为记录名称所对应的数据域
- 许多的行和列组成一张表单
- 若干的表单组成database

非关系型数据库：

NoSQL，指的是非关系型的数据库。NoSQL有时也称作Not Only SQL的缩写，是对不同于传统的关系型数据库的数据库管理系统的统称。NoSQL适用于超大规模数据的存储。（例如谷歌或Facebook每天为他们的用户收集万亿比特的数据）。这些类型的数据存储不需要固定的模式，无需多余操作就可以横向扩展。

类型：

- 键值（Key-Value）存储数据库：键值数据库就类似传统语言中使用哈希表，可以通过key来添加、查询或删除数据，因为使用key主键访问，所以会获得很高的性能及扩展性。典型产品：Memcached、Redis、MemcacheDB、BerkeleyDB 

- 列存储（Column-oriedted）数据库：列存储数据库将数据库存储在列族（column family）中，一个列族存储经常被一起查询的相关数据。举个例子，如果我们有一个Person类，我们通常会一起查询他们的姓名和年龄而不是薪资。这种情况下，姓名和年龄就会被放入一个列族中，而薪资则在另一个列族中。这部分数据库通常用来分布式存储的海量数据，键仍然存在，但是他们的特点是指向了多个列。典型产品：Cassandra、HBase 

- 面向文档（Document-Oriented）数据库：面向文档数据库会将数据以文档的形式存储。每个文档都是自包含的数据单元，是一系列数据项的集合。每个数据项都有一个名称与对应的值，值既可以是简单的数据类型，如字符串、数字和日期等；也可以是复杂的类型，如有序列表和关系对象。数据存储的最小单位是文档，同一个表中存储的文档属性可以是不同的，数据可以使用XML、JSON或者JSONB等多种形式存储。典型产品：MongoDB、CouchDB 

- 图形（Graph）数据库：图形关系的最佳存储。使用传统关系数据库来解决的话性能低下，而且设计使用不方便。典型产品：Neo4J、FlockDB



### MySQL 

MySQL 是最流行的关系型数据库管理系统，在 WEB 应用方面 MySQL 是最好的 RDBMS 应用软件之一。

在我们开始学习MySQL数据库前，让我们先了解下RDBMS的一些术语：

- 数据库: 数据库是一些关联表的集合。
- 数据表: 表是数据的矩阵。在一个数据库中的表看起来像一个简单的电子表格。
- 列: 一列(数据元素) 包含了相同类型的数据, 例如邮政编码的数据。
- 行：一行（=元组，或记录）是一组相关的数据，例如一条用户订阅的数据。
- 冗余：存储两倍数据，冗余降低了性能，但提高了数据的安全性。
- 主键：主键是唯一的。一个数据表中只能包含一个主键。你可以使用主键来查询数据。
- 外键：外键用于关联两个表。
- 复合键：复合键（组合键）将多个列作为一个索引键，一般用于复合索引。
- 索引：使用索引可快速访问数据库表中的特定信息。索引是对数据库表中一列或多列的值进行排序的一种结构。类似于书籍的目录。
- 参照完整性: 参照的完整性要求关系中不允许引用不存在的实体。与实体完整性是关系模型必须满足的完整性约束条件，目的是保证数据的一致性。


#### Download and Installation for MySQL/MariaDB
If in case MySQL is not available, you might use MariaDB instead. 

1. MySQL  USTC 中科大的镜像 http://mirrors.ustc.edu.cn/mysql-ftp/Downloads/ （注意根据你的操作系统选择合适的镜像文件！）

    Windows10 MYSQL Installer 安装 [Click](https://www.runoob.com/w3cnote/windows10-mysql-installer.html)

2. MariaDB 开源软件 可以到其官网下载 https://downloads.mariadb.org/

    Windows10系统下安装MariaDB 的教程图解 [Click](https://pcedu.pconline.com.cn/1367/13676123.html)

**注意：** 在安装过程中要设置root用户的密码，请记住并且妥善保管这个密码，在后续查看及调用数据库的时候都需要使用这个重要的密码。而且root用户是具有最高权限的，不能泄露该密码。

#### 安装pymysql
`pip install pymysql`

### 使用create_engine读写dataframe

The create_engine() function produces an Engine object based on a URL.

`engine = create_engine('dialect+driver://username:password@host:port/database')`
```
dialect：the identifying name of the SQLAlchemy dialect, a name such as sqlite, mysql , postgresql, oracle, or mssql.
driver：the name of the DBAPI to be used to connect to the database using all lowercase letters. Here we use pymysql.
```

In [3]:
from sqlalchemy import create_engine

pw = input('Please provide the password:')
db = input('Please input the name of the database which you would like to access:')

engine = create_engine('mysql+pymysql://root:{}@127.0.0.1:3306/{}?charset=utf8'.format(pw,db))
# 127.0.0.1 代表本机，是特指本机的IP地址，3306是数据库使用的端口号，utf8是编码，一般有中文就应该使用utf8.
# pip install pymysql

Please provide the password: 123456
Please input the name of the database which you would like to access: sakila


**DataFrame.to_sql**(name = SQL table，con = sqlalchemy.engine，schema = None，if_exists = How to behave if the table already exists.)    [Click](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_sql.html)

    if_exists: {‘fail’, ‘replace’, ‘append’}, default ‘fail’.
    - fail: Raise a ValueError.
    - replace: Drop the table before inserting new values.
    - append: Insert new values to the existing table.

**pandas.read_sql**(sql = SQL query or SQL table, con = sqlalchemy.engine)    [Click](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql.html#pandas.read_sql)

In [4]:
import pandas as pd

dat = pd.read_csv('stockdata.csv')
# Write to database
dat.to_sql(name='my_table3', con=engine, if_exists='append', index=False)

In [5]:
sql_cmd = "select * from my_table3 where vol > 100000"

# Read from Database
df = pd.read_sql(sql=sql_cmd, con=engine)
df

trade_date   open   high    low  close  pre_close  change  pct_chg  \
0      20200724  16.80  16.81  15.56  16.20      16.51   -0.31  -1.8776   
1      20200723  17.80  18.07  16.24  16.51      17.95   -1.44  -8.0223   
2      20200722  17.12  18.18  16.90  17.95      17.21    0.74   4.2998   
3      20200721  17.20  17.87  16.77  17.21      16.97    0.24   1.4143   
4      20200720  15.80  16.97  15.50  16.97      15.43    1.54   9.9806   
..          ...    ...    ...    ...    ...        ...     ...      ...   
176    20180402  12.13  12.56  11.80  12.01      12.01    0.00   0.0000   
177    20180330  11.73  12.10  11.50  12.01      11.50    0.51   4.4300   
178    20180329  11.41  12.15  11.39  11.50      11.14    0.36   3.2300   
179    20180327  10.38  11.20  10.38  11.15      10.26    0.89   8.6700   
180    20180207  10.10  10.27   8.94   9.14       9.93   -0.79  -7.9600   

           vol  
0    285487.23  
1    332693.07  
2    184201.74  
3    226621.76  
4    320102.62  
..         ...  
176  176643.38  
177  161507.54  
178  163262.24  
179  117928.51  
180  105255.63  

[181 rows x 9 columns]

利用MySQL Workbench创建数据库

MySQL Workbench使用教程 [Click](http://c.biancheng.net/view/2625.html)

#### Navicat：powerful database management & design GUI 

是一套多连接数据库开发工具，让你在单一应用程序中同时连接多达七种数据库：MySQL、MariaDB、MongoDB、SQL Server、SQLite、Oracle 和 PostgreSQL，可一次快速方便地访问所有数据库。

In [6]:
import pandas as pd
from sqlalchemy import create_engine

dat = pd.read_csv('multi_stock_data.csv')
# Write to database
engine = create_engine('mysql+pymysql://root:123456@127.0.0.1:3306/sakila?charset=utf8')
dat.to_sql(name='my_table', con=engine, if_exists='append', index=False)

In [7]:
sql_cmd = "select * from my_table where ts_code like '%.SH' and close < 100"

df = pd.read_sql(sql=sql_cmd, con=engine)
df.head()

trade_date    ts_code   open    high    low  close  pre_close  change  \
0    20200319  603288.SH  95.50   97.99  92.18  93.40      97.00   -3.60   
1    20200318  603288.SH  96.98   99.35  96.00  97.00      97.00    0.00   
2    20200317  603288.SH  98.00   98.90  92.58  97.00      98.00   -1.00   
3    20200316  603288.SH  98.11  100.97  97.20  98.00     100.98   -2.98   
4    20200228  603288.SH  98.54  102.44  98.54  99.58     102.45   -2.87   

   pct_chg       vol      amount  
0  -3.7113  75266.02  708672.515  
1   0.0000  62069.81  608808.950  
2  -1.0204  80724.36  773597.552  
3  -2.9511  57726.22  570032.586  
4  -2.8014  43809.13  438840.158

#### MySQL 查询数据

**1.MySQL 数据库使用SQL SELECT语句来查询数据。**

以下为在MySQL数据库中查询数据通用的 SELECT 语法：

`SELECT column_name,column_name
FROM table_name
[WHERE Clause]`

- 查询语句中你可以使用一个或者多个表，表之间使用逗号(,)分割，并使用WHERE语句来设定查询条件。
- SELECT 命令可以读取一条或者多条记录。
- 你可以使用星号（*）来代替其他字段，SELECT语句会返回表的所有字段数据
- 你可以使用 WHERE 语句来包含任何条件。

**2.MySQL WHERE 子句**

我们知道从 MySQL 表中使用 SQL SELECT 语句来读取数据，如需有条件地从表中选取数据，可将 WHERE 子句添加到 SELECT 语句中。

以下是 SQL SELECT 语句使用 WHERE 子句从数据表中读取数据的通用语法：

`SELECT field1, field2,...fieldN FROM table_name1, table_name2...
[WHERE condition1 [AND [OR]] condition2.....`

- 查询语句中你可以使用一个或者多个表，表之间使用逗号, 分割，并使用WHERE语句来设定查询条件。
- 你可以在 WHERE 子句中指定任何条件。
- 你可以使用 AND 或者 OR 指定一个或多个条件。
- WHERE 子句也可以运用于 SQL 的 DELETE 或者 UPDATE 命令。
- WHERE 子句类似于程序语言中的 if 条件，根据 MySQL 表中的字段值来读取指定的数据。

**3.MySQL LIKE 子句**

有时我们需要获取某些字段含有特定字符的所有记录，这时我们就需要在 WHERE 子句中使用 SQL LIKE 子句。SQL LIKE 子句中使用百分号 % 字符来表示任意字符，类似于UNIX或正则表达式中的星号 * 。如果没有使用百分号 % , LIKE 子句与等号 = 的效果是一样的。

以下是 SQL SELECT 语句使用 LIKE 子句从数据表中读取数据的通用语法：

`SELECT field1, field2,...fieldN 
FROM table_name
WHERE field1 LIKE condition1 [AND [OR]] filed2 = 'somevalue'`

- 你可以在 WHERE 子句中指定任何条件。
- 你可以在 WHERE 子句中使用LIKE子句。
- 你可以使用LIKE子句代替等号 =。
- LIKE 通常与 % 一同使用，类似于一个元字符的搜索。
- 你可以使用 AND 或者 OR 指定一个或多个条件。
- 你可以在 DELETE 或 UPDATE 命令中使用 WHERE...LIKE 子句来指定条件。

**Exercise：** 

- 练习使用create_engine函数从本地数据库中读写dataframe格式的数据（例如，将文件multi_stock_data.csv存入数据库中）。
- 在练习时尝试使用简单的SQL语句进行查询操作（select from、where、like）

#### 创建数据库连接：pymysql 驱动

In [4]:
import pymysql
 
# 打开数据库连接
db = pymysql.connect("localhost","root","123456")

# 使用 cursor() 方法创建一个游标对象 cursor
cursor = db.cursor()

# 输出所有数据库列表："show databases"
cursor.execute("show databases")
for x in cursor:
    print(x)
    
# # 创建数据库："create database"
# cursor.execute("create database if not exists stock_data")

# # 删除数据库："drop database"
# cursor.execute("drop database if exists stock_data2")

# 关闭数据库连接
db.close()

('information_schema',)
('mysql',)
('performance_schema',)
('sakila',)
('stock_data',)
('sys',)
('world',)


#### 创建、删除表格

In [8]:
import pymysql
 
# 打开数据库连接
db = pymysql.connect("localhost","root","123456","stock_data")
 
# 使用 cursor() 方法创建一个游标对象 cursor
cursor = db.cursor()

# 查看数据表："show tables"
cursor.execute("show tables")
for x in cursor:
    print(x)

# # 创建数据表："create table"
# sql = "create table if not exists EMPLOYEE (NAME CHAR(10) NOT NULL, AGE INT, SEX CHAR(1), INCOME FLOAT)"
# cursor.execute(sql)

# # # 删除数据表：drop table
# sql = "create table if not exists stock_abc (date char(8), price float)"
# cursor.execute(sql)
# cursor.execute("drop table if exists stock_abc")

# 关闭数据库连接
db.close()

('stock_abc',)


#### 插入数据 insert into

In [9]:
import pymysql
 
# 打开数据库连接
db = pymysql.connect("localhost","root","123456","stock_data")
# 使用 cursor() 方法创建一个游标对象 cursor
cursor = db.cursor()

# 创建表
sql = "create table if not exists stocks (stock_code int not null, date char(10), prices float)"
cursor.execute(sql)
print(cursor.rowcount)
# 插入数据
sql = "insert into stocks (stock_code, date, prices) values ('600201', '20200102', '12.3')"
cursor.execute(sql)
db.commit() # 数据表内容有更新，必须使用到该语句
print(cursor.rowcount)

# try:
#     # 执行sql语句
#     cursor.execute(sql)
#     # 提交到数据库执行
#     db.commit()
# except:
#     # 如果发生错误则回滚
#     db.rollback()

# 关闭数据库连接
db.close()

0
1


#### 批量执行 excutemany

In [10]:
import pymysql
 
# 打开数据库连接
db = pymysql.connect("localhost","root","123456","stock_data")
# 使用 cursor() 方法创建一个游标对象 cursor
cursor = db.cursor()

# 插入数据
sql = "insert into stocks (stock_code, date, prices) VALUES (%s, %s, %s)"
val = [(600301, '20200101', 10.1), (600302, '20200101', 10.7), (600303, '20200101', 10.5)]

cursor.executemany(sql, val)
db.commit()
print(cursor.rowcount)

# 关闭数据库连接
db.close()

3


#### 数据获取

Python查询Mysql使用 fetchone() 方法获取单条数据, 使用fetchall() 方法获取多条数据。

- fetchone(): 该方法获取单个查询对象。
- fetchall(): 接收全部的返回结果行.
- rowcount: 这是一个只读属性，返回执行execute()方法后影响的行数。

In [11]:
import pymysql
 
# 打开数据库连接
db = pymysql.connect("localhost","root","123456","stock_data")
# 使用 cursor() 方法创建一个游标对象 cursor
cursor = db.cursor()

sql = "SELECT * FROM stocks"
cursor.execute(sql)

# fetchall()
result = cursor.fetchall()
for x in result:
    print(x)
    
# 关闭数据库连接
db.close()

(600201, '20200102', 12.3)
(600301, '20200101', 10.1)
(600302, '20200101', 10.7)
(600303, '20200101', 10.5)


**Exercise：**

1. 安装pymysql
2. 使用该驱动连接数据库
3. 查看sakila数据库中所有的表的名称。
4. 使用insert into为已有的任意表插入一行，注意数据的格式。

### MongoDB

MongoDB 是一个基于分布式文件存储的数据库。由 C++ 语言编写。旨在为 WEB 应用提供可扩展的高性能数据存储解决方案。

MongoDB 将数据存储为一个文档，数据结构由键值(key=>value)对组成。MongoDB 文档类似于 JSON 对象。字段值可以包含其他文档，数组及文档数组。

Windows 平台安装 MongoDB [Click](https://www.runoob.com/mongodb/mongodb-window-install.html)

#### Installation of pymongo

`pip install pymongo`
 
或者下载相应的wheel文件/tar.gz文件进行安装 

- 安装wheel 先下载文件***.whl， 然后切换到该whl文件的目录。运行 pip install filename.whl 
- 安装tar.gz 先下载文件***.tar.gz, 然后将该文件解压到相应目录。运行 pip install 解压目录  注意该目录中应该有setup.py文件

#### 创建一个数据库

创建数据库需要使用 MongoClient 对象，并且指定连接的 URL 地址和要创建的数据库名。

**注意:** 在 MongoDB 中，数据库只有在内容插入后才会创建! 就是说，数据库创建后要创建集合(数据表)并插入一个文档(记录)，数据库才会真正创建。同样的，集合只有在内容插入后才会创建! 就是说，创建集合(数据表)后要再插入一个文档(记录)，集合才会真正创建。

如下实例中，我们创建的数据库 runoobdb :

In [4]:
import pymongo

#### 连接数据库
myclient = pymongo.MongoClient("mongodb://localhost:27017/")

#### 创建数据库（类似于MySQL中的database）
mydb = myclient["runoobdb"]

#### 创建集合（类似于MySQL中的table）
mycol = mydb["sites"]

#### 在集合中插入文档 - insert_one
mydict = { "name": "RUNOOB", "alexa": "10000", "url": "https://www.runoob.com" }
x = mycol.insert_one(mydict) 

#### 查询一条数据 - fine_one
y = mycol.find_one()
print(y)

mydb

{'_id': ObjectId('5face3c94523aa6f7dd7d993'), 'name': 'RUNOOB', 'alexa': '10000', 'url': 'https://www.runoob.com'}


Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'runoobdb')

#### 判断数据库是否已存在
我们可以读取 MongoDB 中的所有数据库，并判断指定的数据库是否存在：

In [9]:
import pymongo
 
myclient = pymongo.MongoClient('mongodb://localhost:27017/')
 
dblist = myclient.list_database_names()

if "runoobdb" in dblist:
    print("数据库已存在！")
else:
    print("数据库不存在！")

数据库已存在！


#### 判断集合是否已存在

我们可以读取 MongoDB 数据库中的所有集合，并判断指定的集合是否存在：

In [10]:
import pymongo
 
myclient = pymongo.MongoClient('mongodb://localhost:27017/')
 
mydb = myclient['runoobdb']
 
collist = mydb.list_collection_names()

if "sites" in collist:   # 判断 sites 集合是否存在
    print("集合已存在！")

集合已存在！


**Homework（need to be updated before this Saturday night 24:00 pm with jupyternotebook to email liyan_zjgsu@163.com）**

- 在本地创建名为abc的数据库，并创建名为xyz的数据表，将multi_stock_data.csv文件写入该xyz数据表中，读取20200126-20200716这段时间内，股票代码头两位为“00”的数据，并以日期为横坐标，收盘价为纵坐标，绘制折线图，在图中标明横纵坐标，图例在画面左上角，为图片添加网格并设置50%的透明度，线型为点划线，标记为圆圈。

- 结合前几次课的内容，从tushare API中取得10家市值最大的银行行业的上市公司从去年年初至今的股票的日行情数据，并将该数据存入本地名的数据库abc中的表tushare中，存入后在数据库的末尾添加一行均值（无法求均值的列以NaN取代）。从上述10支股票中随机挑选出四支绘制2x2 的子图，横轴为日期，纵轴为收盘价，其中上面两幅子图的横轴不显示；子图1、3为折线图，子2、4为柱状图；为该图添加名为"银行股"的中文标题；在图1、3中，以透明度为80%的蓝色填充大于1.5倍均值的部分。